src/cpp/rtps/transport/shared_mem/SharedMemManager.hpp

```diff
@@ -233,42 +420,57 @@ class SharedMemManager
                 uint32_t size,
                 const std::chrono::steady_clock::time_point& max_blocking_time_point)
         {
-            std::lock_guard<std::mutex> lock(alloc_mutex_);
+            (void)max_blocking_time_point;
 
-            wait_for_avaible_space(size, max_blocking_time_point);
+            std::lock_guard<std::mutex> lock(alloc_mutex_);
 
-            release_unused_buffers();
+            if (!recover_buffers(size))
+            {
+                throw std::runtime_error("allocation overflow");
+            }
 
             void* data = nullptr;
             BufferNode* buffer_node = nullptr;
             std::shared_ptr<SharedMemBuffer> new_buffer;
 
             try
             {
-                buffer_node = static_cast<BufferNode*>(
-                    segment_->get().allocate_aligned(sizeof(BufferNode) + size, 
-                        std::alignment_of<BufferNode>::value));
+                buffer_node = pop_free_node();
 
-                buffer_node->header.data_size = size;
-                buffer_node->header.ref_count.store(0, std::memory_order_relaxed);
+                data = segment_->get().allocate(size);
+                free_bytes_-= size;
+                
+                buffer_node->data_offset = segment_->get_offset_from_address(data);
+                buffer_node->data_size = size;
 
+                auto validity_id = buffer_node->status.load(std::memory_order_relaxed).validity_id;
+                                
                 // TODO(Adolfo) : Dynamic allocation. Use foonathan to convert it to static allocation
-                new_buffer = std::make_shared<SharedMemBuffer>(segment_, segment_id_, buffer_node, segment_node_);
+                new_buffer = std::make_shared<SharedMemBuffer>(segment_, segment_id_, buffer_node, 
+                    static_cast<uint32_t>(validity_id));
 
+                if (new_buffer)
+                {
+                    buffer_node->inc_processing_count(static_cast<uint32_t>(validity_id));
+                }
+                else
+                {
+                    throw std::runtime_error("alloc_buffer: out of memory");
+                }
+                
                 // TODO(Adolfo) : Dynamic allocation. Use foonathan to convert it to static allocation
-                allocated_nodes_.push_back(buffer_node);
-
-                segment_node_->free_bytes.fetch_sub(size);
+                allocated_buffers_.push_back(buffer_node);
```

Comment:
TODO can be removed due to my previous comment.

Reply:
it's a std::list<BufferNode*>, you have to init the pointers to the nodes.